apps/extension/src/background/approvals/service.test.ts

```diff
@@ -0,0 +1,481 @@
+
+      expect(
+        service.submitSignature(tabId, "msgId", signer)
+      ).rejects.toBeDefined();
+    });
+
+    it("should throw an error when data is not present", async () => {
+      const tabId = 1;
+      const signer = "signer";
+      const sigResponse = {
+        hash: "hash",
+        signature: "sig",
+      };
+      (service as any).resolverMap[tabId] = sigResponse;
+      jest.spyOn(dataStore, "get").mockResolvedValueOnce(undefined);
+
+      expect(
+        service.submitSignature(tabId, "msgId", signer)
+      ).rejects.toBeDefined();
+    });
+
+    it("should reject promise if can't sign", async () => {
+      const tabId = 1;
+
+      jest.spyOn(service as any, "getPopupTabId").mockResolvedValue(tabId);
+      jest.spyOn(dataStore, "get").mockResolvedValueOnce("data");
+      jest
+        .spyOn(keyRingService, "signArbitrary")
+        .mockRejectedValue("Can't sign");
+      const signer = "signer";
+
+      await new Promise((resolve) =>
+        setTimeout(() => {
+          resolve(true);
+        })
+      );
+
+      expect(
+        service.submitSignature(tabId, "msgId", signer)
+      ).rejects.toBeDefined();
+    });
+  });
+
+  describe("submitSignature", () => {
+    it("should reject resolver", async () => {
+      const tabId = 1;
+
+      jest.spyOn(service as any, "getPopupTabId").mockResolvedValue(tabId);
+      const signer = "signer";
+      const signaturePromise = service.approveSignature(signer, "data");
+
+      await new Promise((resolve) =>
+        setTimeout(() => {
+          resolve(true);
+        })
+      );
+      await service.rejectSignature(tabId, "msgId");
+
+      expect(signaturePromise).rejects.toEqual(undefined);
+    });
+
+    it("should throw an error if resolver is not found", async () => {
+      const tabId = 1;
+
+      expect(service.rejectSignature(tabId, "msgId")).rejects.toBeDefined();
+    });
+  });
+
+  const txTypes = [
+    [TxType.Bond, "getParamsBond"],
+    [TxType.Unbond, "getParamsUnbond"],
+    [TxType.Withdraw, "getParamsWithdraw"],
+    [TxType.Transfer, "getParamsTransfer"],
+    [TxType.IBCTransfer, "getParamsIbcTransfer"],
+    [TxType.EthBridgeTransfer, "getParamsEthBridgeTransfer"],
+    [TxType.VoteProposal, "getParamsVoteProposal"],
+  ] as const;
+
+  describe("approveTx", () => {
+    it.each(txTypes)("should launch tx", async (type, paramsFn) => {
+      jest.spyOn(ApprovalsService, paramsFn).mockImplementation(() => ({}));
+      jest.spyOn(borsh, "deserialize").mockReturnValue({});
+      jest.spyOn(service as any, "_launchApprovalWindow");
+
+      expect(
+        service.approveTx(type, "", "", AccountType.Mnemonic)
+      ).resolves.not.toBeDefined();
+    });
+  });
+
+  describe("getParamsTransfer", () => {
+    it("should return transfer params", () => {
+      const transferMsgValue = new TransferMsgValue({
+        source: "source",
+        target: "target",
+        token: "token",
+        amount: BigNumber(100),
```

Comment:
That's cool, I didn't know `new` was optional for BigNumber :smile:

Reply:
Me neither :D